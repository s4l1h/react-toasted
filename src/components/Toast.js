import React, { Component } from "react";
import ToastProgress from "./ToastProgress";
import IntervalTimeManager from "../utils/IntervalTimeManager";
import PropsType from "prop-types";
class Toast extends Component {
  constructor(props) {
    super(props);

    this.progressBarTimer = null;
    this.timeoutTimer = null;
    this.state = { percent: 0, showProgressBar: false };
  }
  static displayName = "Toast";
  getPercent = () => {
    if (this.props.progressBarValue !== 0) {
      return this.props.progressBarValue;
    }
    return this.state.percent;
  };
  componentDidMount() {
    if (this.props.onCreated) {
      this.props.onCreated();
    }
    if (this.progressBarTimer != null) {
      this.progressBarTimer.start();
    }
    if (this.timeoutTimer != null) {
      this.timeoutTimer.start();
    }
  }

  componentWillUnmount() {
    if (this.props.onDestroyed) {
      this.props.onDestroyed();
    }
    if (this.progressBarTimer != null) {
      this.progressBarTimer.stop();
    }
    if (this.timeoutTimer != null) {
      this.timeoutTimer.stop();
    }
  }
  componentWillMount() {
    if (this.props.timeout !== 0) {
      // SetUP timeout Manager
      this.timeoutTimer = IntervalTimeManager({
        totalTime: this.props.timeout,
        callbackFunctions: {
          "after:finish": () => {
            if (typeof this.props._remove !== "undefined") {
              this.props._remove(this.props);
            }
            // console.log("Timeout Fired");
          },
          callback: () => {
            this.setState({
              percent: this.progressBarTimer.getPercent()
            });
          }
        }
      });
      // SetUP progressbar Time Manager
      if (this.props.progressBar !== false) {
        this.setState({
          showProgressBar: true
        });
        this.progressBarTimer = IntervalTimeManager({
          totalTime: this.props.timeout
        });
      }
    } else if (
      this.props.progressBar !== false &&
      this.props.progressBarValue !== 0
    ) {
      this.setState({
        showProgressBar: true
      });
    }
  }
  onClick = () => {
    if (this.props.onClick) {
      this.props.onClick();
    }
    if (this.props.removeOnClick) {
      if (typeof this.props._remove !== "undefined") {
        this.props._remove(this.props);
      }
    }
  };
  onMouseOver = () => {
    if (this.progressBarTimer != null) {
      this.progressBarTimer.pause();
    }
    if (this.timeoutTimer != null) {
      this.timeoutTimer.pause();
    }
    if (this.props.onMouseOver) {
      this.props.onMouseOver();
    }
  };
  onMouseOut = () => {
    if (this.progressBarTimer != null) {
      this.progressBarTimer.resume();
    }
    if (this.timeoutTimer != null) {
      this.timeoutTimer.resume();
    }
    if (this.props.onMouseOut) {
      this.props.onMouseOut();
    }
  };
  render() {
    return (
      <div
        style={{
          ...this.props.style
        }}
        className={["toast", "toast-" + this.props.type]
          .concat(this.props.classNames)
          .join(" ")}
        onClick={this.onClick}
        onMouseOver={this.onMouseOver}
        onMouseOut={this.onMouseOut}
      >
        {this.state.showProgressBar && (
          <ToastProgress percent={this.getPercent()} />
        )}
        {this.props.title !== "" && (
          <div className="toast-title">{this.props.title}</div>
        )}
        <div className="toast-message">{this.props.msg}</div>
      </div>
    );
  }
}
Toast.defaultProps = {
  type: "success",
  title: "",
  msg: "",
  classNames: [],
  style: {},
  removeOnClick: true,
  onClick: false,
  onMouseOver: false,
  onMouseOut: false,
  onCreated: false,
  onDestroyed: false,
  timeout: 5000,
  progressBar: true,
  progressBarValue: 0
};
Toast.propTypes = {
  /** Toast Title */
  title: PropsType.string,
  /** Toast Message */
  msg: PropsType.string.isRequired,
  /** Toast Injected Class Names */
  classNames: PropsType.array,
  /** Toast Injected Style */
  style: PropsType.object,
  /** Toast Position */
  position: PropsType.string.isRequired,
  /** Toast Type */
  type: PropsType.oneOf(["success", "warning", "error", "info"]).isRequired,
  /* Close Toast When it clicked */
  removeOnClick: PropsType.bool,
  /** Toast onClick event */
  onClick: PropsType.oneOfType([PropsType.bool, PropsType.func]),
  /** Toast onMouseOver event function */
  onMouseOver: PropsType.oneOfType([PropsType.bool, PropsType.func]),
  /** Toast onMouseOut event function */
  onMouseOut: PropsType.oneOfType([PropsType.bool, PropsType.func]),
  /** Toast onCreated event function */
  onCreated: PropsType.oneOfType([PropsType.bool, PropsType.func]),
  /** Toast onDestroyed event function */
  onDestroyed: PropsType.oneOfType([PropsType.bool, PropsType.func]),
  /** Toast Timeout time */
  timeout: PropsType.number,
  /** Show Progress Bar? */
  progressBar: PropsType.bool,
  /** Static Progress Bar Value */
  progressBarValue: PropsType.number
};
export default Toast;
