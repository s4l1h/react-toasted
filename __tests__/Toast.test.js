// __tests__/hidden-message.js
// these imports are something you'd normally configure Jest to import for you
// automatically. Learn more in the setup docs: https://testing-library.com/docs/react-testing-library/setup#cleanup
// import "@testing-library/react/cleanup-after-each";
import "@testing-library/jest-dom/extend-expect";

// NOTE: jest-dom adds handy assertions to Jest and is recommended, but not required

import React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react";
import Toast from "../src/components/Toast";

describe("Toast Component Snapshot", () => {
  test("basic toast snapshot", () => {
    const { container } = render(<Toast position="toast-top-left" />);
    expect(container).toMatchSnapshot();
  });
  test("it has to match snapshot", () => {
    const { container } = render(
      <Toast
        {...{
          name: "SampleFull",
          position: "toast-bottom-full-width",
          title: "Full",
          type: "success",
          msg: "Look at the console",
          removeOnClick: false,
          classNames: ["animated"],
          style: { color: "red" },
          timeout: 10000,
          progressBar: true,
          progressBarValue: 0
        }}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
describe("Toast Events", () => {
  let toastMessage = "Message";
  let onCreated = jest.fn();
  let onClicked = jest.fn();
  let onMouseOver = jest.fn();
  let onMouseOut = jest.fn();
  let onDestroyed = jest.fn();
  const { getByText, unmount } = render(
    <Toast
      msg={toastMessage}
      position="toast-top-right"
      onCreated={onCreated}
      onClick={onClicked}
      onMouseOver={onMouseOver}
      onMouseOut={onMouseOut}
      onDestroyed={onDestroyed}
    />
  );
  test("onCreated", () => {
    expect(onCreated).toHaveBeenCalledTimes(1);
  });
  test("onClicked", () => {
    fireEvent.click(getByText(toastMessage));
    expect(onClicked).toHaveBeenCalledTimes(1);
  });

  test("mouseOver", () => {
    fireEvent.mouseOver(getByText(toastMessage));
    expect(onMouseOver).toBeCalled();
  });

  test("onMouseOut", () => {
    fireEvent.mouseOut(getByText(toastMessage));
    expect(onMouseOut).toBeCalled();
  });
  test("onDestroyed", () => {
    unmount();
    expect(onDestroyed).toBeCalled();
    expect(onDestroyed).toHaveBeenCalledTimes(1);
  });
});
