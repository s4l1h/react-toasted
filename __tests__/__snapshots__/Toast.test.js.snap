// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast Component Snapshot basic toast snapshot 1`] = `
<div>
  <div
    class="toast toast-success"
  >
    <div
      class="toast-message"
    />
  </div>
</div>
`;

exports[`Toast Component Snapshot it has to match snapshot 1`] = `
<div>
  <div
    class="toast toast-success animated"
    style="color: red;"
  >
    <div
      class="toast-title"
    >
      Full
    </div>
    <div
      class="toast-message"
    >
      Look at the console
    </div>
  </div>
</div>
`;
